README.md
dbs2
====

Goals

 - [x] read upgrade files from configured directory
 - [x] sort files into x.y.z release number order
 - [x] break upgrade file contents into XX.YY schema revisions
 - [x] determine what revisions need to be executed
 - [x] execute revisions & update tracker
 - [x] built in cli script
 - [x] configuration loaded from json file
 - [x] custom configuration via script
 - [x] Runner Commands (tests, updates, downgrades, etc) use response object to track status and messaging

TODO
 - [x] Switch to a specific CliRunner class
 - [x] CliRunner class sets verbosity level
 - [x] Runner class figures out what to do
   - [x] connection test
   - [x] DbSmart2 table test
   - [x] RevisionCheck
   - [x] Run Updates
   - [ ] Run Downgrades
   - [ ] Dump Upgrade Log
 - [x] Runner Class tests database connection
 - [x] Runner Class tests for DbSmart2 table
 - [x] Runner Class can setup DbSmart2 table
 - [x] Runner Class checks which revisions need to be run
 - [x] Runner Class runs updates
 - [x] Runner Class updates tracker table
 - [x] CliRunner class handles all output responsibilities
 - [ ] Changelog
 - [ ] Documentation


CHANGELOG.md
# CHANGELOG

## [1.0.1] - 2016-Oct-28

### Fixed
- Uncaught PDOException: schema_date doesn't have a default value (#1)

[1.0.1]: https://github.com/mike27cubes/dbs2/compare/1.0.0...1.0.1


tests/unit/ConfigTest.php
<?php

/**
 * Tests the DbSmart2 Config class
 *
 * PHP version 5.3
 *
 * @vendor     27 Cubes
 * @package    DbSmart2
 * @subpackage DbSmart2Tests
 * @author     27 Cubes <info@27cubes.net>
 * @since      1.0.0
 */

namespace Cubes\DbSmart2Tests;

class ConfigTest extends \PHPUnit_Framework_TestCase
{
    /**
     * Useless test to start out
     */
    public function testGetInstance()
    {
        $obj = new \Cubes\DbSmart2\Config();
        $this->assertInstanceOf('\Cubes\DbSmart2\Config', $obj);
    }

    public function testInitialStateGetters()
    {
        $obj = new \Cubes\DbSmart2\Config();
        $this->assertEmpty($obj->getDsn());
        $this->assertEmpty($obj->getDbName());
        $this->assertEmpty($obj->getUsername());
        $this->assertEmpty($obj->getPassword());
        $this->assertEmpty($obj->getUpgradePath());
    }

    public function testConstructorLoadGetters()
    {
        $dbname = uniqid('dbname');
        $input = array(
            'dsn' => uniqid('mysql:dbname=' . $dbname . ';host=127.0.0.1'),
            'username' => uniqid('username-'),
            'password' => uniqid('password-'),
            'upgrade_path' => __DIR__
        );
        $obj = new \Cubes\DbSmart2\Config($input);
        $this->assertEquals($input['dsn'], $obj->getDsn());
        $this->assertEquals($dbname, $obj->getDbName());
        $this->assertEquals($input['username'], $obj->getUsername());
        $this->assertEquals($input['password'], $obj->getPassword());
        $this->assertEquals($input['upgrade_path'], $obj->getUpgradePath());
    }

    /**
     * @expectedException BadMethodCallException
     */
    public function testCallInvalidMethodName()
    {
        $obj = new \Cubes\DbSmart2\Config();
        $obj->someInvalidMethod();
    }

    /**
     * @expectedException BadMethodCallException
     */
    public function testGetterInvalidMethodName()
    {
        $obj = new \Cubes\DbSmart2\Config();
        $obj->getInvalidMethod();
    }

    /**
     * @expectedException InvalidArgumentException
     */
    public function testLoadFromJsonFile_FileNotFound()
    {
        $obj = new \Cubes\DbSmart2\Config();
        $obj->loadFromJsonFile(__DIR__ . '/file_not_found.json');
    }

    public function testLoadFromJsonFileGetters()
    {
        $obj = new \Cubes\DbSmart2\Config();
        $filepath = dirname(__DIR__) . '/data/test-config.json';
        $retVal = $obj->loadFromJsonFile($filepath);
        $this->assertEquals($obj, $retVal);
        $expected = json_decode(trim(file_get_contents($filepath)), true);
        $this->assertEquals($expected['dsn'], $obj->getDsn());
        $this->assertEquals($expected['username'], $obj->getUsername());
        $this->assertEquals($expected['password'], $obj->getPassword());
        $this->assertEquals(str_replace('__DIR__', dirname($filepath), $expected['upgrade_path']), $obj->getUpgradePath());
    }

    public function testLoadFromJsonFile_CustomConfig()
    {
        $obj = new \Cubes\DbSmart2\Config();
        $configroot = dirname(__DIR__) . '/data/';
        $configdata = $configroot . 'test-config.json';
        $configfile = $configroot . 'test-config-customloader.json';
        $retVal = $obj->loadFromJsonFile($configfile);
        $this->assertEquals($obj, $retVal);
        $expected = json_decode(trim(file_get_contents($configdata)), true);
        $this->assertEquals($expected['dsn'], $obj->getDsn());
        $this->assertEquals($expected['username'], $obj->getUsername());
        $this->assertEquals($expected['password'], $obj->getPassword());
        // Note: Now replacements done on custom loaded config data
        $this->assertEquals($expected['upgrade_path'], $obj->getUpgradePath());
    }
}


tests/unit/RunnerTest.php
<?php

/**
 * Tests the DbSmart2 Runner class
 *
 * PHP version 5.3
 *
 * @vendor     27 Cubes
 * @package    DbSmart2
 * @subpackage DbSmart2Tests
 * @author     27 Cubes <info@27cubes.net>
 * @since      1.0.0
 */

namespace Cubes\DbSmart2Tests;

class RunnerTest extends \PHPUnit_Framework_TestCase
{
    /**
     * Useless test to start out
     */
    public function testGetInstance()
    {
        $obj = new \Cubes\DbSmart2\Runner();
        $this->assertInstanceOf('\Cubes\DbSmart2\Runner', $obj);
    }

    public function testLoadConfig()
    {
        $obj = new \Cubes\DbSmart2\Runner();
        $ret = $obj->loadConfig(new \Cubes\DbSmart2\Config(array('upgrade_path' => dirname(__DIR__) . '/data/runner')));
        $this->assertInstanceOf('\Cubes\DbSmart2\Runner', $ret);
        $this->assertEquals($obj, $ret);
    }

    /**
     * @expectedException RuntimeException
     */
    public function testGetQueryList_NoConfigLoaded()
    {
        $obj = new \Cubes\DbSmart2\Runner();
        $obj->getQueryList();
    }

    public function testGetQueryList()
    {
        $obj = new \Cubes\DbSmart2\Runner();
        $list = $obj->loadConfig(new \Cubes\DbSmart2\Config(array('upgrade_path' => dirname(__DIR__) . '/data/runner')))
                    ->getQueryList();
        $expected = new \Cubes\DbSmart2\QueryList(array(
            'x1' => array(
                'y1' => 'SELECT CURDATE();',
                'y2' => 'SELECT CURTIME();'
            ),
            'x2' => array(
                'y3' => 'SELECT NOW();',
            ),
            'x3' => array(
                'y4' => 'SELECT UNIX_TIMESTAMP();',
            )
        ));
        $this->assertEquals($expected, $list);
    }
}


tests/phpunitbootstrap.php
<?php

/**
 * Composer autoload
 */
require_once dirname(__DIR__) . '/vendor/autoload.php';


library/Runner.php
<?php

/**
 * DbSmart2 Runner
 *
 * PHP version 5.3
 *
 * @vendor     27 Cubes
 * @package    DbSmart2
 * @author     27 Cubes <info@27cubes.net>
 * @since      1.0.0
 */

namespace Cubes\DbSmart2;

/**
 * Runner
 *
 * @vendor     27 Cubes
 * @package    DbSmart2
 * @since      1.0.0
 */
class Runner
{
    /**
     * Configuration
     *
     * @var Config
     */
    protected $config;

    /**
     * DB Connection
     *
     * @var \PDO
     */
    protected $db;

    const COMMAND_REVISIONCHECK = 'revisioncheck';
    const COMMAND_UPGRADE = 'upgrade';
    const COMMAND_DOWNGRADE = 'downgrade';
    const COMMAND_CONNECTIONTEST = 'connectiontest';
    const COMMAND_TABLETEST = 'tabletest';
    const COMMAND_SETUPTRACKER = 'setuptracker';
    const COMMAND_DUMPLOG = 'dumplog';
    const COMMAND_NULL = 'null';

    const TRACKER_TABLE = 'DbSmart2';

    /**
     * Sets the config
     *
     * @param  Config $config
     * @return Runner fluent interface
     */
    public function loadConfig(Config $config)
    {
        $this->config = $config;
        return $this;
    }

    /**
     * Gets the query list
     *
     * @return QueryList
     * @throws \RuntimeException
     */
    public function getQueryList()
    {
        if (!($this->config instanceof Config)) {
            throw new \RuntimeException('Config not loaded');
        }
        $lister = new UpgradeFileLister();
        $loader = new UpgradeFileLoader();
        return $loader->loadFiles($lister->getFileList($this->config));
    }

    /**
     * Gets a list of the "new" schema ids
     *
     * @return array
     */
    public function getNewSchemaIds()
    {
        $queries = $this->getQueryList();
        $queries->filterOutQueries($this->getExecutedSchemaIds());
        return $queries->getSchemaIds();
    }

    /**
     * Runs the given command
     *
     * @param  string $command the command to run
     * @param  array  $options
     * @return Response
     */
    public function run($command, $options = array())
    {
        $runMethod = 'run' . ucfirst($command);
        $response = new Response();
        return call_user_func(array($this, $runMethod), $options, $response);
    }

    /**
     * Run Revision Check
     *
     * @param  array    $options
     * @param  Response $response
     * @return Response
     */
    protected function runRevisioncheck($options = array(), Response $response)
    {
        $response = $this->runConnectiontest($options, $response);
        if ($response->hasFailures()) {
            return $response;
        }
        $response = $this->runTabletest($options, $response);
        if ($response->hasFailures()) {
            return $response;
        }
        $new = $this->getNewSchemaIds();
        if (count($new) > 0) {
            $response->addResult(self::COMMAND_REVISIONCHECK, false, 'Revisions to run' . "\n\n" . join("\n", $new) . "\n");
        } else {
            $response->addResult(self::COMMAND_REVISIONCHECK, true, 'Up To Date');
        }
        return $response;
    }

    /**
     * Run Downgrade
     *
     * @param  array $options
     * @param  Response $response
     * @return Response
     */
    protected function runDowngrade($options = array(), Response $response)
    {
        throw new \RuntimeException(__METHOD__ . ' not implemented');
    }

    /**
     * Run Connection Test
     *
     * @param  array $options
     * @param  Response $response
     * @return Response
     */
    protected function runConnectiontest($options = array(), Response $response)
    {
        try {
            $db = $this->getDb();
            $response->addResult(self::COMMAND_CONNECTIONTEST, true, 'Connection Test Passed');
        } catch (\Exception $e) {
            $response->addResult(self::COMMAND_CONNECTIONTEST, false, $e->getMessage());
        }
        return $response;
    }

    /**
     * Run Table Test
     *
     * @param  array $options
     * @param  Response $response
     * @return Response
     */
    protected function runTabletest($options = array(), Response $response)
    {
        $db = $this->getDb();
        $sql = 'SHOW CREATE TABLE ' . self::TRACKER_TABLE;
        $exists = false;
        try {
            foreach ($db->query($sql) as $row) {
                if (!empty($row)) {
                    $exists = true;
                }
            }
            if ($exists) {
                $response->addResult(self::COMMAND_TABLETEST, true, 'Tracker Table exists');
            } else {
                $response->addResult(self::COMMAND_TABLETEST, false, 'Tracker Table does not exist');
            }
        } catch (\PDOException $e) {
            $response->addResult(self::COMMAND_TABLETEST, false, 'Tracker Table does not exist: ' . $e->getMessage());
        }
        return $response;
    }

    /**
     * Run Setup Tracker
     *
     * @param  array $options
     * @param  Response $response
     * @return Response
     */
    protected function runSetuptracker($options = array(), Response $response)
    {
        $resposne = $this->runConnectiontest($options, $response);
        if ($response->hasFailures()) {
            return $response;
        }
        $db = $this->getDb();
        $sql = 'CREATE TABLE ' . self::TRACKER_TABLE . ' (
            schema_id VARCHAR ( 16 ) NOT NULL ,
            schema_md5 VARCHAR ( 32 ) NOT NULL ,
            schema_date DATE NOT NULL ,
            schema_time TIME NOT NULL ,
            PRIMARY KEY ( schema_id ) ,
            INDEX ( schema_md5 )
        ) ENGINE=InnoDb DEFAULT CHARSET=UTF8';
        try {
            $db->exec($sql);
            $response->addResult(self::COMMAND_SETUPTRACKER, true, 'Tracker Table Created');
        } catch (\PDOException $e) {
            $response->addResult(self::COMMAND_SETUPTRACKER, false, 'Tracker Table creation failed: ' . $e->getMessage());
        }
        return $response;
    }

    /**
     * Run Dump Log
     *
     * @param  array $options
     * @param  Response $response
     * @return Response
     */
    protected function runDumplog($options = array(), Response $response)
    {
        throw new \RuntimeException(__METHOD__ . ' not implemented');
    }

    /**
     * Run Null Command
     *
     * For Testing Purposes only
     *
     * @param  array $options
     * @param  Response $response
     * @return Response
     */
    protected function runNull($options = array(), Response $response)
    {
        return $response->addResult(self::COMMAND_NULL, true, trim('NULL ' . json_encode($options)));
    }

    /**
     * Run Upgrade
     *
     * @param  array $options
     * @param  Response $response
     * @return Response
     */
    public function runUpgrade($options = array(), Response $response)
    {
        $queries = $this->getQueryList();
        $queries->filterOutQueries($this->getExecutedSchemaIds());
        $db = $this->getDb();
        foreach ($queries->getQueries() as $querySetId => $querySet) {
            foreach ($querySet as $queryId => $queryBlock) {
                // TODO: Start transaction
                $splitQueries = $this->splitUpQueryBlock($queryBlock);
                foreach ($splitQueries as $query) {
                    $stmt = $db->prepare($query);
                    if ($stmt === false) {
                        // TODO: ROLLBACK
                        return $response->addResult(self::COMMAND_UPGRADE, false, 'Could not prepare query: ' . join(' - ', $db->errorInfo()));
                    }
                    if (!$stmt->execute()) {
                        // TODO: ROLLBACK
                        return $response->addResult(self::COMMAND_UPGRADE, false, 'Could not execute query: ' . join(' - ', $stmt->errorInfo()));
                    }
                    $stmt->closeCursor();
                }
                if (!$this->logExecutedSchemaId($querySetId, $queryId, $queryBlock)) {
                    // TODO: ROLLBACK
                    return $response->addResult(self::COMMAND_UPGRADE, false, 'Could not update tracker table');
                }
                // TODO: Commit
            }
        }
        return $response->addResult(self::COMMAND_UPGRADE, true, 'Successfully executed queries' . "\n" . join("\n", $queries->getSchemaIds()));
    }

    /**
     * Logs the given query-set-id and query-id to the tracker table
     *
     * @param  string $querySetId
     * @param  string $queryId
     * @param  string $queryBlock
     * @return bool
     */
    protected function logExecutedSchemaId($querySetId, $queryId, $queryBlock)
    {
        $sql = 'INSERT INTO DbSmart2 (schema_id, schema_md5, schema_date, schema_time) VALUES(?, ?, CURDATE(), CURTIME())';
        $db = $this->getDb();
        $stmt = $db->prepare($sql);
        if ($stmt === false) {
            return false;
        }
        return $stmt->execute(array($querySetId . '.' . $queryId, md5(trim($queryBlock))));
    }

    /**
     * Splits up the given query block into a list of executable queries
     *
     * @param  string $queryBlock
     * @return array
     */
    protected function splitUpQueryBlock($queryBlock)
    {
        $queries = explode("\n", trim($queryBlock));
        foreach ($queries as $k => $v) {
            $queries[$k] = rtrim($v);
        }
        $queryBlock = join("\n", $queries);
        return explode(";\n", $queryBlock);
    }

    /**
     * Gets a list of the executed schema ids
     *
     * @return array
     */
    protected function getExecutedSchemaIds()
    {
        $db = $this->getDb();
        $stmt = $db->prepare('SELECT schema_id FROM DbSmart2 ORDER BY schema_id ASC');
        if ($stmt === false) {
            throw new \RuntimeException('Could not poll executed schema ids: ' . join(' - ', $db->errorInfo()));
        }
        if (!$stmt->execute()) {
            throw new \RuntimeException('Could not poll executed schema ids: ' . join(' - ', $stmt->errorInfo()));
        }
        return $stmt->fetchAll(\PDO::FETCH_COLUMN);
    }

    /**
     * Gets the Datbase Connection
     *
     * @return \PDO
     * @throws \RuntimeException, \PDOException
     */
    public function getDb()
    {
        if ($this->db instanceof \PDO) {
            return $this->db;
        }
        if (!($this->config instanceof Config)) {
            throw new \RuntimeException('Config not loaded');
        }
        $db = new \PDO($this->config->getDsn(), $this->config->getUsername(), $this->config->getPassword(), array(\PDO::ATTR_ERRMODE => \PDO::ERRMODE_EXCEPTION));
        $this->db = $db;
        return $this->db;
    }

    /**
     * Validates the given command
     *
     * @param  string $toTest
     * @return bool
     */
    public function validateCommand($toTest)
    {
        return in_array($toTest, self::getCommandList());
    }

    /**
     * Gets a list ov valid commands
     *
     * @return array
     */
    public static function getCommandList()
    {
        return array(
            self::COMMAND_REVISIONCHECK,
            self::COMMAND_UPGRADE,
            self::COMMAND_DOWNGRADE,
            self::COMMAND_CONNECTIONTEST,
            self::COMMAND_TABLETEST,
            self::COMMAND_SETUPTRACKER,
            self::COMMAND_DUMPLOG,
            self::COMMAND_NULL
        );
    }
}


library/Response.php
<?php

/**
 * DbSmart2 Response
 *
 * PHP version 5.3
 *
 * @vendor     27 Cubes
 * @package    DbSmart2
 * @author     27 Cubes <info@27cubes.net>
 * @since      1.0.0
 */

namespace Cubes\DbSmart2;

/**
 * Response
 *
 * @vendor     27 Cubes
 * @package    DbSmart2
 * @since      1.0.0
 */
class Response
{
    /**
     * Results
     *
     * @var array
     */
    protected $results = array();

    /**
     * Adds a result to the response
     *
     * @param  string $command
     * @param  bool   $status  Whether the result represents a successful action
     * @param  string $message
     * @return self   Fluent interface
     */
    public function addResult($command, $status, $message)
    {
        $this->results[] = array('command' => $command, 'status' => $status, 'message' => $message);
        return $this;
    }

    /**
     * Gets the results
     *
     * @return array
     */
    public function getResults()
    {
        return $this->results;
    }

    /**
     * Determines if the response contains failures
     *
     * @return bool
     */
    public function hasFailures()
    {
        foreach ($this->results as $result) {
            if ($result['status'] === false) {
                return true;
            }
        }
        return false;
    }
}


library/CliRunner.php
<?php

/**
 * DbSmart2 CLI Runner
 *
 * PHP version 5.3
 *
 * @vendor     27 Cubes
 * @package    DbSmart2
 * @author     27 Cubes <info@27cubes.net>
 * @since      1.0.0
 */

namespace Cubes\DbSmart2;

/**
 * CLI Runner
 *
 * @vendor     27 Cubes
 * @package    DbSmart2
 * @since      1.0.0
 */
class CliRunner
{
    /**
     * Options
     *
     * @var array
     */
    protected $options = array();

    /**
     * Verbosity level
     *
     * @var int
     */
    protected $verbosity = self::VERBOSITY_NORMAL;

    const VERBOSITY_QUIET = 0;
    const VERBOSITY_NORMAL = 1;
    const VERBOSITY_LOUD = 2;

    /**
     * Constructor
     *
     * @param  array $options
     */
    public function __construct($options = array())
    {
        $this->options = $options;
        $this->setupVerbosity();
    }

    /**
     * Setup verbosity level based on options
     *
     * @return CliRunner fluent interface
     */
    protected function setupVerbosity()
    {
        $this->verbosity = self::VERBOSITY_NORMAL; // default verbosity level
        if (!empty($this->options['V'])) {
            $this->verbosity = self::VERBOSITY_LOUD;
        }
        if (!empty($this->options['q'])) {
            $this->verbosity = self::VERBOSITY_QUIET;
        }
        return $this;
    }

    /**
     * Determines if the verbosity level is "loud"
     *
     * @return bool
     */
    protected function loudOutput()
    {
        return $this->verbosity == self::VERBOSITY_LOUD;
    }

    /**
     * Determines if the verbosity level is "normal" or greater
     *
     * @return bool
     */
    protected function normalOutput()
    {
        return in_array($this->verbosity, array(self::VERBOSITY_NORMAL, self::VERBOSITY_LOUD));
    }

    /**
     * Run
     *
     * @param  string $projectRoot
     * @param  string $command
     * @param  array  $commandOptions
     */
    public function run($projectRoot, $command, $commandOptions = array())
    {
        if (!empty($this->options['h'])) {
            return $this->runHelp();
        }
        $command = strtolower($command);
        $runner = new Runner();
        if (!$runner->validateCommand($command)) {
            return $this->runHelp('Invalid Command "' . $command . '"');
        }
        // Load Config
        $configPath = $projectRoot . '/dbsmart2.json';
        $config = new Config();
        $config->loadFromJsonFile($configPath);
        $response = $runner->loadConfig($config)
                           ->run($command, $commandOptions);
        if ($this->loudOutput()) {
            foreach ($response->getResults() as $result) {
                echo 'Command: ' . $result['command'] . "\n";
                echo 'Status: ' . $result['status'] . "\n";
                echo 'Message: ' . $result['message'] . "\n\n";
            }
        } elseif ($this->normalOutput()) {
            foreach ($response->getResults() as $result) {
                echo $result['message'] . "\n";
            }
        } else {
            // Assumed quiet
            $failures = array();
            if ($response->hasFailures()) {
                foreach ($response->getResults() as $result) {
                    if ($result['status'] === false) {
                        $failures[] = $result['command'] . ': ' . $result['message'];
                    }
                }
                $stdErr = fopen('php://stderr', 'a');
                fwrite($stdErr, "\n" . join("\n", $failures) . "\n");
                fclose($stdErr);
            }
        }
    }

    /**
     * Runs the Help Command
     *
     * @param  string $error
     */
    public function runHelp($error = '')
    {
        echo <<<'EOH'
DBSMART2

Usage: project/vendor/bin/dbsmart2 [switches] [command]

Switches
  -h  View Help Message (what you are viewing right now)
  -v  Normal verbosity. Prints all command results and success messages to
      STDOUT, errors to STDERR.
  -V  Loud verbosity. Prints all command results, success messages and debug
      info to STDOUT, errors to STDERR.
  -q  Quiet mode. Nothing is printed to STDOUT, cancels out -v and -V. Errors
      are still printed to STDERR.

Commands
  revisioncheck  - Check to see which revisions need to be executed. This is the
                   default command. Automatically executed by upgrade command.
  upgrade        - Execute all pending revisions.
  setuptracker   - Sets up the DBSMART tracker table.
  connectiontest - Test the configured database connection. Automatically
                   executed by revisioncheck, upgrade, setuptracker, tabletest
                   and dumplog commands.
  tabletest      - Test for the existence of the DBSMART tracker table.
                   Automatically executed by revisioncheck, upgrade,
                   setuptracker, and dumplog commands.
  dumplog        - Dump the contents of the DBSMART tracker table.

Proposed Commands
  downgrade      - Downgrade by executing specific single or sets of rollback
                   revisions

EOH;
        if (!empty($error) && is_scalar($error)) {
            trigger_error('DBSMART2 ERROR' . "\n" . $error, E_USER_ERROR);
        }
    }
}


library/QueryList.php
<?php

/**
 * DbSmart2 Query List
 *
 * PHP version 5.3
 *
 * @vendor     27 Cubes
 * @package    DbSmart2
 * @author     27 Cubes <info@27cubes.net>
 * @since      1.0.0
 */

namespace Cubes\DbSmart2;

/**
 * Query List
 *
 * @vendor     27 Cubes
 * @package    DbSmart2
 * @since      1.0.0
 */
class QueryList
{
    /**
     * The queries
     *
     * @var array
     */
    protected $queries = array();

    /**
     * Constructor
     *
     * @param  array $queries [optional] the queries
     */
    public function __construct($queries = array())
    {
        if (is_array($queries) && !empty($queries)) {
            // TODO: validate structure of array eg $queries[xx][yy] = string
            $this->queries = $queries;
        }
    }

    /**
     * Gets the queries from this list
     *
     * @return array
     */
    public function getQueries()
    {
        return $this->queries;
    }

    /**
     * Gets the schema ids
     *
     * @return array
     */
    public function getSchemaIds()
    {
        $ids = array();
        foreach ($this->queries as $xx => $grouped) {
            foreach ($grouped as $yy => $ignored) {
                $ids[] = $xx . '.' . $yy;
            }
        }
        return $ids;
    }

    /**
     * Joins the given query list into
     *
     * @param  QueryList $toJoin
     * @return QueryList fluent interface
     */
    public function joinWith(QueryList $toJoin)
    {
        foreach ($toJoin->getQueries() as $xx => $grouped) {
            foreach ($grouped as $yy => $queryBlock) {
                if (!isset($this->queries[$xx])) {
                    $this->queries[$xx] = array();
                }
                if (!isset($this->queries[$xx][$yy])) {
                    $this->queries[$xx][$yy] = '';
                }
                $this->queries[$xx][$yy] = trim($this->queries[$xx][$yy] . "\n" . $queryBlock);
            }
        }
        return $this;
    }

    /**
     * Filters out queries
     *
     * @param  array $idsToRemove
     * @return QueryList
     */
    public function filterOutQueries($idsToRemove = array())
    {
        // Remove values that don't match the DBSmart Schema Id pattern
        $idsToRemove = array_filter($idsToRemove, function($v) {
            return strpos($v, '.') !== false;
        });
        foreach ($idsToRemove as $schemaId) {
            list($xx, $yy) = explode('.', $schemaId, 2);
            if (isset($this->queries[$xx][$yy])) {
                unset($this->queries[$xx][$yy]);
            }
        }
        foreach (array_keys($this->queries) as $xx) {
            $this->queries[$xx] = array_filter($this->queries[$xx]);
        }
        return $this;
    }
}


library/UpgradeFileLister.php
<?php

/**
 * DbSmart2 Upgrade File Lister
 *
 * PHP version 5.3
 *
 * @vendor     27 Cubes
 * @package    DbSmart2
 * @author     27 Cubes <info@27cubes.net>
 * @since      1.0.0
 */

namespace Cubes\DbSmart2;

/**
 * Upgrade File Lister
 *
 * @vendor     27 Cubes
 * @package    DbSmart2
 * @since      1.0.0
 */
class UpgradeFileLister
{
    /**
     * Gets the Upgrade File List
     *
     * @param  \Cubes\DbSmart2\Config $config
     * @return array
     */
    public function getFileList(Config $config)
    {
        $files = array_unique(glob($config->getUpgradePath() . '/*.sql'));
        $files = array_filter($files, array($this, 'isFileRelevant'));
        usort($files, array($this, 'upgradeFileSortComparison'));
        return $files;
    }

    /**
     * Determines if the given file is relevant
     *
     * @param  string $file
     * @return bool
     */
    public function isFileRelevant($file)
    {
        $name = trim(substr($file, strrpos($file, '/')), '/');
        if ($name == 'branch-upgrade.sql' || preg_match('/upgrade-([0-9]+)_([0-9]+)_([0-9]+).sql/', $name)) {
            return true;
        }
        return false;
    }

    /**
     * Compares the upgrade files in order to sort the upgrade file list
     *
     * @param  string $a
     * @param  string $b
     * @return int -1,0,1
     */
    public function upgradeFileSortComparison($a, $b)
    {
        $aName = trim(substr($a, strrpos($a, '/')), '/');
        $bName = trim(substr($b, strrpos($b, '/')), '/');
        if ($aName == 'branch-upgrade.sql') {
            return 1;
        } elseif ($bName == 'branch-upgrade.sql') {
            return -1;
        }
        $aVersion = trim(substr($aName, strrpos($aName, '-')), '-');
        $aVersion = array_pad(explode('_', trim(substr($aVersion, 0, strpos($aVersion, '.')), '.')), -3, 0);
        $bVersion = trim(substr($bName, strrpos($bName, '-')), '-');
        $bVersion = array_pad(explode('_', trim(substr($bVersion, 0, strpos($bVersion, '.')), '.')), -3, 0);
        for ($i = 0; $i < 3; $i++) {
            if ((int) $aVersion[$i] > (int) $bVersion[$i]) {
                return 1;
            } elseif ((int) $aVersion[$i] < (int) $bVersion[$i]) {
                return -1;
            }
        }
        return 0;
    }
}


library/Config.php
<?php

/**
 * DbSmart2 Config
 *
 * PHP version 5.3
 *
 * @vendor     27 Cubes
 * @package    DbSmart2
 * @author     27 Cubes <info@27cubes.net>
 * @since      1.0.0
 */

namespace Cubes\DbSmart2;

/**
 * Config
 *
 * @vendor     27 Cubes
 * @package    DbSmart2
 * @since      1.0.0
 */
class Config
{
    /**
     * DSN
     *
     * @var string
     */
    protected $dsn;

    /**
     * DB Username
     *
     * @var string
     */
    protected $username;

    /**
     * DB Password
     *
     * @var string
     */
    protected $password;

    /**
     * Path to upgrade files
     *
     * @var string
     */
    protected $upgradePath;

    /**
     * Config constructor
     *
     * @param  array $details
     */
    public function __construct($details = array())
    {
        if (is_array($details) && count($details) > 0) {
            $this->loadConfig($details);
        }
    }

    /**
     * Loads the configuration from the given json file
     *
     * @param  string $filePath
     * @return Config fluent interface
     * @throws \InvalidArgumentException
     */
    public function loadFromJsonFile($filePath)
    {
        if (!file_exists($filePath)) {
            throw new \InvalidArgumentException('Config file "' . $filePath . '" does not exist');
        }
        $contents = trim(file_get_contents($filePath));
        return $this->loadConfig(json_decode($this->parseFileReplacements($contents, $filePath), true));
    }

    /**
     * Parses some replacements from the config file
     *
     * @param  string $contents config file contents
     * @param  string $filePath path to config file
     * @return string parsed file contents
     */
    protected function parseFileReplacements($contents, $filePath)
    {
        $replacements = array(
            '__DIR__' => dirname($filePath)
        );
        foreach ($replacements as $search => $replace) {
            $contents = str_replace($search, $replace, $contents);
        }
        return $contents;
    }

    /**
     * Loads the config details
     *
     * @param  array $details
     * @return Config fluent interface
     */
    public function loadConfig($details = array())
    {
        if (!empty($details['customconfig'])) {
            return $this->loadCustomConfig($details['customconfig']);
        }
        $map = array(
            'dsn' => 'dsn',
            'username' => 'username',
            'password' => 'password',
            'upgrade_path' => 'upgradePath'
        );
        foreach ($map as $param => $property) {
            if (isset($details[$param])) {
                $this->$property = $details[$param];
            }
        }
        return $this;
    }

    /**
     * Loads config via a custom script
     *
     * Script must return an array of data readable by Config::loadConfig()
     *
     * @param  string $configscript path to custom configuration script
     * @return Config fluent interface
     */
    protected function loadCustomConfig($configscript)
    {
        return $this->loadConfig(require $configscript);
    }

    /**
     * Gets the configured Database Name
     *
     * @return string
     */
    public function getDbName()
    {
        $matches = array();
        // mysql:dbname=testdb;host=127.0.0.1
        if (!preg_match('/dbname=([a-zA-Z0-9\._-]+);/', $this->getDsn(), $matches)) {
            return '';
        }
        return $matches[1];
    }

    /**
     * Magic Method caller
     *
     * @param  string $name
     * @param  array  $params
     * @return mixed
     * @throws \BadMethodCallException
     */
    public function __call($name, $params = array())
    {
        if (substr($name, 0, 3) == 'get') {
            return $this->getter($name, $params);
        }
        throw new \BadMethodCallException('Invalid method "' . $name . '"');
    }

    /**
     * Magic Method Getter
     *
     * @param  string $name
     * @param  array  $params
     * @return mixed
     * @throws \BadMethodCallException
     */
    protected function getter($name, $params = array())
    {
        $property = lcfirst(substr($name, 3));
        if (!property_exists($this, $property)) {
            throw new \BadMethodCallException('Invalid getter "' . $name . '"');
        }
        return $this->$property;
    }
}
